explorer/interpreter/type_checker.cpp

```diff
@@ -954,199 +982,268 @@ class ConstraintTypeBuilder {
+  // constraint's self binding. The `self_witness` is the witness for the
+  // resulting constraint, and can be `GetSelfWitness()`. The `bindings`
+  // parameter specifies any additional substitutions to perform.
+  void AddAndSubstitute(const TypeChecker& type_checker,
+                        Nonnull<const ConstraintType*> constraint,
+                        Nonnull<const Value*> self,
+                        Nonnull<const Witness*> self_witness,
+                        const Bindings& bindings, bool add_lookup_contexts) {
+    // First substitute into the impl bindings to form the full witness for
+    // the constraint type.
+    std::vector<Nonnull<const Witness*>> witnesses;
     for (const auto& impl_constraint : constraint->impl_constraints()) {
-      AddImplConstraint(impl_constraint);
-    }
+      Bindings local_bindings = bindings;
+      local_bindings.Add(constraint->self_binding(), self,
+                         type_checker.MakeConstraintWitness(
+                             *constraint, witnesses,
+                             constraint->self_binding()->source_loc()));
+      int index = AddImplConstraint(
+          {.type =
+               type_checker.Substitute(local_bindings, impl_constraint.type),
+           .interface = cast<InterfaceType>(type_checker.Substitute(
+               local_bindings, impl_constraint.interface))});
+      witnesses.push_back(
+          type_checker.MakeConstraintWitnessAccess(self_witness, index));
+    }
+
+    // Now form a complete witness and substitute it into the rest of the
+    // constraint.
+    Bindings local_bindings = bindings;
+    local_bindings.Add(constraint->self_binding(), self,
+                       type_checker.MakeConstraintWitness(
+                           *constraint, std::move(witnesses),
+                           constraint->self_binding()->source_loc()));
 
     for (const auto& equality_constraint : constraint->equality_constraints()) {
-      AddEqualityConstraint(equality_constraint);
+      std::vector<Nonnull<const Value*>> values;
+      for (const Value* value : equality_constraint.values) {
+        // Ensure we don't create any duplicates through substitution.
+        if (std::find_if(values.begin(), values.end(), [&](const Value* v) {
+              return ValueEqual(v, value, std::nullopt);
+            }) == values.end()) {
+          values.push_back(type_checker.Substitute(local_bindings, value));
+        }
+      }
+      AddEqualityConstraint({.values = std::move(values)});
     }
 
-    for (const auto& lookup_context : constraint->lookup_contexts()) {
-      AddLookupContext(lookup_context);
+    if (add_lookup_contexts) {
+      for (const auto& lookup_context : constraint->lookup_contexts()) {
+        AddLookupContext({.context = type_checker.Substitute(
+                              local_bindings, lookup_context.context)});
+      }
     }
   }
 
+  // Bring all the `impl`s accumulated so far into the given impl scope.
```

Comment:
We're not being great on this here, but I will note "Function comments should be written with an implied subject of _This function_ and should start with the verb phrase; for example, "Opens the file", rather than "Open the file"."

https://google.github.io/styleguide/cppguide.html#Function_Comments

Bring -> Brings, Add -> Adds, Set up -> Sets up, etc

Reply:
Done, throughout this class at least.